connect-inject/cleanup_resource_test.go

```diff
@@ -0,0 +1,322 @@
+			}
+
+			// Create the cleanup resource.
+			log := hclog.Default().Named("cleanupResource")
+			log.SetLevel(hclog.Debug)
+			consulURL, err := url.Parse("http://" + server.HTTPAddr)
+			require.NoError(err)
+			kubeResources := c.KubePods
+			if !c.OutOfClusterNode {
+				node := nodeName(t, consulClient)
+				// NOTE: we need to add the node because the reconciler checks if
+				// the node the service is registered with actually exists in this
+				// cluster.
+				kubeResources = append(kubeResources, &corev1.Node{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: node,
+					},
+				})
+
+			}
+			cleanupResource := CleanupResource{
+				Log:              log,
+				KubernetesClient: fake.NewSimpleClientset(kubeResources...),
+				ConsulClient:     consulClient,
+				ConsulScheme:     consulURL.Scheme,
+				ConsulPort:       consulURL.Port(),
+			}
+
+			// Run Reconcile.
+			cleanupResource.reconcile()
+
+			// Test that the remaining services are what we expect.
+			services, err := consulClient.Agent().Services()
+			require.NoError(err)
+			var actualServiceIDs []string
+			for id := range services {
+				actualServiceIDs = append(actualServiceIDs, id)
+			}
+			require.ElementsMatch(actualServiceIDs, c.ExpConsulServiceIDs)
+		})
+	}
+}
+
+func TestDelete(t *testing.T) {
+	t.Parallel()
+
+	cases := map[string]struct {
+		Pod                 *corev1.Pod
+		ConsulServices      []capi.AgentServiceRegistration
+		ExpConsulServiceIDs []string
+		ExpErr              string
+	}{
+		"pod is nil": {
+			Pod:    nil,
+			ExpErr: "object for key default/foo was nil",
+		},
+		"pod does not have service-name annotation": {
+			Pod: &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "foo-abc123",
+					Namespace: "default",
+				},
+				Status: corev1.PodStatus{
+					HostIP: "127.0.0.1",
+				},
+			},
+			ExpErr: "pod did not have consul.hashicorp.com/connect-service annotation",
+		},
+		"no instances still registered": {
+			Pod:                 fooPod,
+			ConsulServices:      nil,
+			ExpConsulServiceIDs: nil,
+		},
+		"app and sidecar terminated": {
+			Pod:                 fooPod,
+			ConsulServices:      []capi.AgentServiceRegistration{consulFooSvc, consulFooSvcSidecar},
+			ExpConsulServiceIDs: nil,
+		},
+		"only app is registered, no sidecar": {
+			Pod:                 fooPod,
+			ConsulServices:      []capi.AgentServiceRegistration{consulFooSvc},
+			ExpConsulServiceIDs: nil,
+		},
+		"only sidecar is registered, no app": {
+			Pod:                 fooPod,
+			ConsulServices:      []capi.AgentServiceRegistration{consulFooSvcSidecar},
+			ExpConsulServiceIDs: nil,
+		},
+		"multiple instances of the same service": {
+			Pod: fooPod,
+			ConsulServices: []capi.AgentServiceRegistration{
+				consulFooSvc,
+				consulFooSvcSidecar,
+				consulFooSvcPod2,
+				consulFooSvcSidecarPod2,
+			},
+			ExpConsulServiceIDs: []string{"foo-def456-foo", "foo-def456-foo-sidecar-proxy"},
+		},
```

Comment:
Do we need tests that check that consul service instances without pod and k8s ns meta keys are ignored?

Reply:
Yep good catch, added:
```go
		"instance does not have pod-name meta": {
			Pod: fooPod,
			ConsulServices: []capi.AgentServiceRegistration{consulNoPodNameMetaSvc},
			ExpConsulServiceIDs: []string{"foo-abc123-foo"},
		},
		"instance does not have k8s-namespace meta": {
			Pod: fooPod,
			ConsulServices: []capi.AgentServiceRegistration{consulNoK8sNSMetaSvc},
			ExpConsulServiceIDs: []string{"foo-abc123-foo"},
		},

```